dohproxy/server_protocol.py

```diff
@@ -28,17 +29,75 @@ class DOHDNSException(DOHException):
     pass
 
 
-class DNSClientProtocol:
-    def __init__(self, dnsq, queue, clientip, logger=None):
+class DNSClient():
+
+    DEFAULT_TIMEOUT = 10
+
+    def __init__(self, upstream_resolver, upstream_port):
+        self.loop = asyncio.get_event_loop()
+        self.upstream_resolver = upstream_resolver
+        self.upstream_port = upstream_port
+
+    async def query(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
+        dnsr = await self.query_udp(dnsq, clientip, timeout=timeout)
+        if dnsr is None or (dnsr.flags & dns.flags.TC):
+            dnsr = await self.query_tcp(dnsq, clientip, timeout=timeout)
+        return dnsr
+
+    async def query_udp(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
+        qid = dnsq.id
+        fut = asyncio.Future()
+        await self.loop.create_datagram_endpoint(
+            lambda: DNSClientProtocolUDP(dnsq, fut, clientip),
+            remote_addr=(self.upstream_resolver, self.upstream_port))
+        return await self._try_query(fut, qid, timeout)
+
+    async def query_tcp(self, dnsq, clientip, timeout=DEFAULT_TIMEOUT):
+        qid = dnsq.id
+        fut = asyncio.Future()
+        await self.loop.create_connection(
+            lambda: DNSClientProtocolTCP(dnsq, fut, clientip),
+            self.upstream_resolver, self.upstream_port)
+        return await self._try_query(fut, qid, timeout)
+
+    async def _try_query(self, fut, qid, timeout):
+        try:
+            await asyncio.wait_for(fut, timeout)
+            dnsr = fut.result()
+            dnsr.id = qid
+        except asyncio.TimeoutError:
+            self.logger.debug("Request timed out")
```

Comment:
this will fail because no logger is defined in this class. Handle this like in DNSClientProtocol